### PR TITLE
Ensure tilt angles array is not sized too large

### PR DESCRIPTION
**Pull request by psavery**

### CODE REVIEW

**Comment by psavery** [comment]
This fixes the issue I mention in #1751.

@cryos, what do you think of this fix?

**Comment by cryos** [comment]
This doesn't make sense to me, we would never want to resize the tilt angles array to such a large number, where is that number coming from? Isn't vtkIdType a 64 bit unsigned int, and you are then comparing to a signed 32 bit int. What number is being generated, and where is it coming from?

File: tomviz/DataSource.cxx [hunk 1]
```diff
@@ -42,18 +42,22 @@
...
 namespace {
-void createOrResizeTiltAnglesArray(vtkDataObject* data)
+void createOrResizeTiltAnglesArray(
+  vtkDataObject* data,
+  vtkIdType maxSize = std::numeric_limits<vtkIdType>::max())
 {
   auto fd = data->GetFieldData();
   if (!fd->HasArray("tilt_angles")) {
     int* extent = vtkImageData::SafeDownCast(data)->GetExtent();
     int numTiltAngles = extent[5] - extent[4] + 1;
```
Comment (cryos) [inline tomviz/DataSource.cxx]: I am guessing one of these numbers is coming out off, any realistic data set should have 1 - maybe a few thousand projections.

**Comment by psavery** [comment]
@cryos Sorry, I submitted this PR in a rush before I left the office. I should have written some explanation to go with it.

What's happening in the issue is that, when the angles get set in the EMD reader [here](https://github.com/OpenChemistry/tomviz/blob/master/tomviz/EmdFormat.cxx#L664), the allocation for the angle array that occurs [here](https://github.com/OpenChemistry/tomviz/blob/master/tomviz/DataSource.cxx#L1168) is too large (for the tilt series example, the allocation is for 255 elements, when the number of tilt angles is only 74). That is because the allocation is made based upon the extent [here](https://github.com/OpenChemistry/tomviz/blob/master/tomviz/DataSource.cxx#L53) rather than the number of angles given in `setTiltAngles()`. This resulted in a lot of tilt angles of `0` after the first 74 tilt angles, so I wanted to provide a way for `setTiltAngles()` to communicate to `createOrResizeTiltAnglesArray()` how much it should allocate, that's what the `maxSize` parameter is for.

If `maxSize` was set by the caller, it would cause the angle array to allocate that many elements (74 in our case). And if `maxSize` was left to its default state, it would have no impact on the function (because `numTiltAngles < maxSize` would always be true). I used vtkIdType here because I looked up the type that `SetNumberOfTuples()` takes (I didn't actually notice that `numTiltAngles` was an int; I could change `maxSize` and `numTuples` to be ints instead).

So yeah, basically this PR was an attempt to provide a way to supply `createOrResizeTiltAnglesArray()` with the actual array size that was needed, so it wouldn't allocate too much space for the array.

An alternative one-line fix for this bug is to put `tiltAngles->SetNumberOfTuples(angles.size());` right after [this line](https://github.com/OpenChemistry/tomviz/blob/master/tomviz/DataSource.cxx#L1170). The only downside is that an array too large will usually be generated, and then it will be sized down immediately after. I could do this fix instead, let me know.

**Comment by cryos** [comment]
createOrResizeTiltAnglesArray is a private function, setting std::numeric_limits<>::max() makes no sense, and the type used is larger than what can be accommodated by vtkImageData for extents (signed int). The number of slices in z should always match the number of tilt angles, I would like to understand why the vtkImageData passed in from EmdFormat is giving 255 for extents when it looks like it should have the correct number of slices.

We should be doing a sanity check, probably in the reader and/or the static set tilt angles method, to confirm the number of angles is equal to the number of slices in z before attempting to set them. I don't have time to look at this right now, but adding some of these sanity checks would be better.

**Comment by psavery** [comment]
@cryos Okay. I think I have a much better fix for this. I'll submit the PR later tonight.

**Comment by cryos** [comment]
It can wait until Monday, but I think I see the real issue as I was just taking a look - line 681 in EmdFormat.cxx performs the permute to swap axes long after the angles have been set. That block just needs moving up before the angles are set so that the number of slices is the same as the number of angles. If they ever differ there are likely to be other issues.

**Comment by cryos** [comment]
PR #1832 works well here with your new tilt series. The logic was a little odd, I think this is simpler but could still do with a little more in future.

**Comment by psavery** [comment]
@cryos Thanks for taking a look at this! Yeah, that change in the EMD reader was similar to what I was going to try this evening. Sorry I didn't notice this issue sooner.

**Pull request closed without merging**